src/ConnectedMode/QualityProfiles/QualityProfileDownloader.cs

```diff
@@ -0,0 +1,202 @@
+            ISonarQubeService sonarQubeService,
+            IBindingConfigProvider bindingConfigProvider,
+            IConfigurationPersister configurationPersister,
+            ISolutionBindingOperation solutionBindingOperation,
+            ILogger logger,
+            IEnumerable<Language> languagesToBind)
+        {
+            this.bindingConfigProvider = bindingConfigProvider;
+            this.sonarQubeService = sonarQubeService;
+            this.configurationPersister = configurationPersister;
+            this.solutionBindingOperation = solutionBindingOperation;
+            this.logger = logger;
+            this.languagesToBind = languagesToBind;
+        }
+
+        public async Task<bool> UpdateAsync(BoundSonarQubeProject boundProject, IProgress<FixedStepsProgress> progress, CancellationToken cancellationToken)
+        {
+            // TODO - CancellableJobRunner
+            // TODO - threading
+            // TODO - skip downloading up to date QPs
+
+            EnsureProfilesExistForAllSupportedLanguages(boundProject);
+
+            var bindingConfigs = new List<IBindingConfig>();
+
+            var languageCount = languagesToBind.Count();
+            int currentLanguage = 0;
+
+            foreach (var language in languagesToBind)
+            {
+                currentLanguage++;
+
+                var progressMessage = string.Format(BindingStrings.DownloadingQualityProfileProgressMessage, language.Name);
+                progress?.Report(new FixedStepsProgress(progressMessage, currentLanguage, languageCount));
+
+                var qualityProfileInfo = await TryDownloadQualityProfileAsync(boundProject, language, cancellationToken);
+
+                if (qualityProfileInfo == null)
+                {
+                    continue; // skip to the next language
+                }
+                UpdateProfile(boundProject, language, qualityProfileInfo);
+
+                var bindingConfiguration = configurationPersister.Persist(boundProject);
+
+                // Create the binding configuration for the language
+                var bindingConfig = await bindingConfigProvider.GetConfigurationAsync(qualityProfileInfo, language, bindingConfiguration, cancellationToken);
+                if (bindingConfig == null)
+                {
+                    logger.WriteLine(string.Format(BindingStrings.SubTextPaddingFormat,
+                        string.Format(BindingStrings.FailedToCreateBindingConfigForLanguage, language.Name)));
+                    return false;
+                }
+
+                bindingConfigs.Add(bindingConfig);
+
+                logger.WriteLine(string.Format(BindingStrings.SubTextPaddingFormat,
+                    string.Format(BindingStrings.QualityProfileDownloadSuccessfulMessageFormat, qualityProfileInfo.Name, qualityProfileInfo.Key, language.Name)));
+            }
+
+            solutionBindingOperation.SaveRuleConfiguration(bindingConfigs, cancellationToken);
```

Comment:
Line 133 does what the now-redundant `IBindingProcessImpl.SaveRuleConfiguration` method did i.e. write the actual rule configuration files to disc.